compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -210,19 +226,49 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
         NestedVisitorMap::None
     }
 
+    fn visit_arm(&mut self, arm: &'tcx Arm<'tcx>) {
+        if arm.guard.is_some() {
+            self.nested_scope_of_guards.push(<_>::default());
+            self.arm_has_guard = true;
+        }
+        self.visit_pat(&arm.pat);
+        if let Some(ref g) = arm.guard {
+            match g {
+                Guard::If(ref e) => {
+                    self.visit_expr(e);
+                }
+            }
+            let mut scope_var_ids =
+                self.nested_scope_of_guards.pop().expect("should have pushed at least one earlier");
+            for var_id in scope_var_ids.drain(..) {
+                assert!(
+                    self.current_scope_of_guards.remove(&var_id),
+                    "variable should be placed in scope earlier"
+                );
+            }
+            self.arm_has_guard = false;
```

Comment:
We should reset back to the original value, which won't always be false (if there were a match nested inside the guard). Here's a (very contrived) example..

```
match x {
  Some(y) if match y.len() { 0 => false, l if l % 2 == 0 => true, _ => false } => (),
  _ => (),
}
```

In fact, I'm tempted to say this flag is redundant and we should remove it. We can query `current_scope_of_guards` every time, and only find a way to optimize that if it's slow.